### PR TITLE
[Core] fix erase iterator while iterating over a map.

### LINKED ISSUE #17080: [placeholder] Investigate Dask on Ray segfault from dlmalloc in nightly release tests
https://buildkite.com/ray-project/periodic-ci/builds/555#_

### PR DESCRIPTION
**Pull request by scv119**
erasing an [absl](https://github.com/abseil/abseil-cpp/blob/master/absl/container/internal/raw_hash_set.h#L1248)/[std](https://en.cppreference.com/w/cpp/container/map/erase) map iterator invalidates the iterator. 

this leads to SIGABRT when I running debug build wheels. ([one of the crash stack got from core dump](https://gist.github.com/scv119/db8e9837dcb45c3abd86f551e29f0871))

instead, we should use _Post-increment_  i.e., ```map.erase(iterator++)``` for absl::flat_hash_map, where

> Post-increment and post-decrement creates a copy of the object, increments or decrements the value of the object and returns the copy from before the increment or decrement.

https://en.cppreference.com/w/cpp/language/operator_incdec

basically ```map.erase(iterator++)``` equals to
```
auto copy = iterator;
iterator++;
map.erase(copy);
```
Note as @clarkzinzow points out 
```
it = map.erase(it)
```
is more idiomatic code, but unfortunately absl::flat_has_map.erase returns void :(

### CODE REVIEW

**Review by rkooo567 — APPROVED** [review]

**Review by ericl — APPROVED** [review]

**Comment by scv119** [comment]
let me sweep through the codebase to see if there are more of this case...

**Review by clarkzinzow — APPROVED** [review]
Great catch! I think that we should look at adding a [ubsan](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html) Buildkite job to our CI, if that seems feasible.

**Comment by scv119** [comment]
@clarkzinzow  yup! I think running debug build, ubsa/asan/tsan build would be super useful. The debug build have already found a few potential deadlock issues running integration tests. https://buildkite.com/ray-project/ray-builders-branch/builds/2408#47f37fa5-73d5-490e-86ef-ffff7d01fa0c

**Comment by scv119** [comment]
src/ray/gcs/gcs_server/gcs_placement_group_manager.cc suffers from similar problem but more tricky.  this [trace](https://gist.github.com/scv119/0a1aad65a05e345412cd90b595126ae3) shows the crash running `pytest ray/tests/test_placement_group.py -k test_automatic_cleanup_job`

**Comment by rkooo567** [comment]
What’s the overhead of tsan / ubsan? (Also are they subset of other *san build or they are completely orthogona?) we should start adding them from next sprint to ci or release tests

File: src/ray/object_manager/plasma/create_request_queue.cc [hunk 1]
```diff
@@ -190,9 +190,10 @@ void CreateRequestQueue::RemoveDisconnectedClientRequests(
 
   for (auto it = fulfilled_requests_.begin(); it != fulfilled_requests_.end();) {
     if (it->second && it->second->client == client) {
-      fulfilled_requests_.erase(it);
+      fulfilled_requests_.erase(it++);
```
Comment (clarkzinzow) [inline src/ray/object_manager/plasma/create_request_queue.cc:193]: Actually, I'd push for using the returned iterator here, since that's considered the idiomatic pattern (even included in the `map.erase()` docs as an example):
```suggestion
      it = fulfilled_requests_.erase(it);
```

**Review by clarkzinzow — DISMISSED** [review]

File: src/ray/object_manager/plasma/create_request_queue.cc (hunk 1 above)
Comment (scv119) [reply src/ray/object_manager/plasma/create_request_queue.cc:193]: yup `it = fulfilled_requests_.erase(it);` definitely much better code. 
However, it's unfortunate that this absl map.[erase returns void](https://github.com/abseil/abseil-cpp/blob/master/absl/container/flat_hash_map.h#L223) :'(

**Review by stephanie-wang — APPROVED** [review]

File: src/ray/object_manager/plasma/create_request_queue.cc (hunk 1 above)
Comment (clarkzinzow) [reply src/ray/object_manager/plasma/create_request_queue.cc:193]: Ah I forgot that this is an `absl::flat_hash_map`!

**Review by clarkzinzow — APPROVED** [review]
